renderer: store used buffers per monitor

gulafaran opened the pull request:
instead of storing used async bufs in the renderer and dropping them on the first monitor pageflip, store them per monitor and drop on the monitors pageflip.

problem before was the first monitor could be dropping the second monitors surfaces before it had even rendered them causing glitches. especially noticeable on dual gpu laptop and external monitor.

move surface frames to CSurfaceManager aswell and schedule them for scanout, and make third pending frame call it aswell in new_render_scheduling.

#TODO
- [ ] ensure the new_render_scheduler actually works as intended
- [x] test toplevel export

fixes: #12181
maybe: #10979

vaxerski reviewed (commented):
yes, but we need a cross-monitor solution too...

I propose to make a manager like `CBufferReleaseManager` that has:
 - `::add(surf, mon)`
 - `::drop(mon)`

Each surface stored gets a `state` with a `std::vector<PHLMONITORREF>` and it gets dropped once that vec has either 0 elements or all elements are null or disabled (in case of a monitor disconnect)

gulafaran commented:
> yes, but we need a cross-monitor solution too...
> 
> I propose to make a manager like `CBufferReleaseManager` that has:
> 
>     * `::add(surf, mon)`
> 
>     * `::drop(mon)`
> 
> 
> Each surface stored gets a `state` with a `std::vector<PHLMONITORREF>` and it gets dropped once that vec has either 0 elements or all elements are null or disabled (in case of a monitor disconnect)

can be a way to deal with surfaces showing on more then one monitor yes, should buffers even be dropped if the rendermode isnt RENDER_MODE_NORMAL?

gulafaran commented:
@vaxerski did some pseudo thing here. it breaks DS here but rest seems fine. and im a bit unsure if on that visible check and also the CHLBufferReference implentation, is it like a shared pointer or do the monitors just end up with 2 different local ones so the syncreleasers are seperate and doesnt actually merge the fences of both, but just keeps them seperate one per monitor.

but this is in general is the idea that should probably be done. to get buffer releases somewhat right. new_render_shed i got no clue if it broke even further. it already was here heh

gulafaran commented:
@vaxerski ok got buffer releases working without breaking DS or OBS now. i still dont know how to proceed with new_render_shed tho since its all sorts of wonky even before

gulafaran commented:
@Zebra2711 wanna test this now? i still have to figure out if there can be something done on laptops tho, seems now its blitting a lot more making it actually slower on new_render_shed when having dualgpu/external monitor tho

gulafaran commented:
@outfoxxed seeing you had various stall/lag issues would be neat if you could test aswell.

what i do know is DS might require some finicking now that commitpendinganddoexplicit is called on new_render_scheduling while it isnt on setting it false.

gulafaran commented:
> > @Zebra2711 wanna test this now? i still have to figure out if there can be something done on laptops tho, seems now its blitting a lot more making it actually slower on new_render_shed when having dualgpu/external monitor tho
> 
> It hit my performance so hard, my hyprland on the iGPU couldn't get more than 60 fps with or without `new_render_scheduling` enabled

unless ofc you just measured something that is vsynced, glxgears is normally. i dont think it has been correct before. hyprland pushes like 100fps if i crunch hard on the gpu before glxgears would still report 240., well if we are pageflipping at 100fps we cant exactly render glxgears at 240.

Zebra2711 commented:
> Lol wat in what? I finally got overlay showing 240 + 160 here and cs2 200+ on external

idk, my hyprland build with #11000, #11087, #11228, and #11216, and your old closed PR #11303, with `new_render_scheduling=true` when started. Everything is extremely smooth 🙃

gulafaran commented:
new_render_sched also hits the blitting harder in AQ if this is a dual gpu thing. and its very slow sadly, needs some work.

> started. Everything is extremely

what setup is it, dual gpu? dual monitors?

Zebra2711 commented:
ok, I built hyprland with #11000, #11087, #11228, and #11216. This MR seems to be performant, close to #11303, and I tested it with `glxgears`.
186xx fps (#11303) vs 172xx fps this MR. `debug:overlay = false`

> what setup is it, dual gpu? dual monitors?

intel igpu + nvidia gpu and no external display

Zebra2711 commented:
i think the main cause of the lag here is that hyprland is using more gpu than it normally would. Lag is reduced compared to the old version, but not much. I still feel laggy with this. Could be that my config is messing with this. I will test it out with the default config tomorrow

Out of the scope of this, when I resize a window or move my mouse, the igpu usage just jumps up

outfoxxed commented:
> @outfoxxed seeing you had various stall/lag issues would be neat if you could test aswell.
> 
> what i do know is DS might require some finicking now that commitpendinganddoexplicit is called on new_render_scheduling while it isnt on setting it false.

I can try to test it soon

gulafaran committed:
renderer: store used buffers per monitor

instead of storing used async bufs in the renderer and dropping them on
the first monitor pageflip, store them per monitor and schedule drop on
the proper fence in the new CBufferReleaseManager.

gulafaran committed:
renderer: make new_render_scheduler not drop buffers wrong

make new_render_scheduler not drop buffers to early when third frame is
pending, by adding a parameter to commitPendingAndDoExplicitSync when to
drop buffers. also reuse monitor inFence to reduce flushing.

gulafaran committed:
renderer: remove copy paste

remove copy paste return early from debugging.

gulafaran committed:
renderer: add damage to second monitor

if the buffer gets locked to another monitor add the intersecting region
to monitor damage and schedule a frame.

gulafaran committed:
renderer: avoid races with commited frame

if we have pending scanout a frame might be waiting on the sync and
racing the buffer drops from said commit. try to avoid that from
occuring.

gulafaran committed:
renderer: schedule frame callbacks

move BufferReleaseManager to SurfaceManager and schedule frame callbacks
and send them on presented and ahead of third monitor rendering in new
frame sheduler.

gulafaran committed:
surfacemgr: also remove monitor from scheduled frames

remove monitor from scheduled frames on destroy.

gulafaran committed:
surfacemgr: schedule all frames

schedule all frames instead of directly sending them.

kRHYME7 commented:
Hi! How should I test this one, any logs that can help?

gulafaran commented:
> Hi! How should I test this one, any logs that can help?

nah thats fine i sort of know whats going wrong, there is two issues one is frame callbacks is being sent wrong. and explicit sync breaks, i just havent managed to cleanly implent it without breaking other things. so i have just put this on hold for now
